### PR TITLE
Move input values into an array (#870)

### PR DESCRIPTION
**Pull request by jroweboy**
As @neobrain suggested in #870 I moved the input mappings into an array and use an enum of Key Values to index the array.

### CODE REVIEW

**Comment by chinhodado** [comment]
The point of making an array is so that you can iterate through the keys without having to repeat things. So you probably will also need an array of default keys for GLFW, an array of default keys for Qt, which will probably lead to an array of PadState as well.

File: src/core/settings.h [hunk 1]
```diff
@@ -8,31 +8,19 @@
...
-    int pad_dright_key;
-    int pad_sup_key;
-    int pad_sdown_key;
-    int pad_sleft_key;
-    int pad_sright_key;
-    int pad_cup_key;
-    int pad_cdown_key;
-    int pad_cleft_key;
-    int pad_cright_key;
+    int input_mappings[NativeInput::NUM_INPUTS];
```
Comment (lioncash) [inline src/core/settings.h]: should use a std::array for this.

File: src/core/settings.h [hunk 2]
```diff
@@ -8,31 +8,19 @@
 
 namespace Settings {
 
+enum NativeInput {
```
Comment (neobrain) [inline src/core/settings.h]: *enum class NativeInput

**Comment by neobrain** [comment]
As @chinhodado, this is only the first step, and existing code should furthermore be changed to use a simple iteration rather than duplicating code.

File: src/core/settings.h (hunk 2 above)
Comment (Subv) [inline src/core/settings.h]: You can't use members of an enum class directly as array indices, you'd have to add a static_cast for each usage.

Comment (chinhodado) [inline src/core/settings.h]: ...which is cumbersome, and defeats the whole purpose of using this enum mostly for array indexing.

Comment (neobrain) [inline src/core/settings.h]: You can add a casting operator to an integer type. That's not particularly clean, but it works and keeps the namespace sorted (being able to access `A` through `Settings::A` is really odd).

Comment (yuriks) [inline src/core/settings.h]: Meh, there's no point. I'd rather just add a prefix to the enum members, or stick it inside its own namespace.

**Comment by jroweboy** [comment]
I've updated the PR to use for loops as requested

File: src/citra/config.cpp [hunk 3]
```diff
@@ -39,31 +39,21 @@ bool Config::LoadINI(INIReader* config, const char* location, const std::string&
...
-    Settings::values.pad_dright_key = glfw_config->GetInteger("Controls", "pad_dright", GLFW_KEY_H);
-    Settings::values.pad_sup_key    = glfw_config->GetInteger("Controls", "pad_sup",    GLFW_KEY_UP);
-    Settings::values.pad_sdown_key  = glfw_config->GetInteger("Controls", "pad_sdown",  GLFW_KEY_DOWN);
-    Settings::values.pad_sleft_key  = glfw_config->GetInteger("Controls", "pad_sleft",  GLFW_KEY_LEFT);
-    Settings::values.pad_sright_key = glfw_config->GetInteger("Controls", "pad_sright", GLFW_KEY_RIGHT);
-    Settings::values.pad_cup_key    = glfw_config->GetInteger("Controls", "pad_cup",    GLFW_KEY_I);
-    Settings::values.pad_cdown_key  = glfw_config->GetInteger("Controls", "pad_cdown",  GLFW_KEY_K);
-    Settings::values.pad_cleft_key  = glfw_config->GetInteger("Controls", "pad_cleft",  GLFW_KEY_J);
-    Settings::values.pad_cright_key = glfw_config->GetInteger("Controls", "pad_cright", GLFW_KEY_L);
+    for (int i=0; i < Settings::NativeInput::NUM_INPUTS; ++i) {
```
Comment (turbedi) [inline src/citra/config.cpp]: missing space before and after `=`

File: src/citra/emu_window/emu_window_glfw.cpp [hunk 4]
```diff
@@ -138,32 +138,9 @@ void EmuWindow_GLFW::DoneCurrent() {
...
-
-    KeyMap::SetKeyMapping({Settings::values.pad_cright_key, keyboard_id}, Service::HID::PAD_C_RIGHT);
-    KeyMap::SetKeyMapping({Settings::values.pad_cleft_key,  keyboard_id}, Service::HID::PAD_C_LEFT);
-    KeyMap::SetKeyMapping({Settings::values.pad_cup_key,    keyboard_id}, Service::HID::PAD_C_UP);
-    KeyMap::SetKeyMapping({Settings::values.pad_cdown_key,  keyboard_id}, Service::HID::PAD_C_DOWN);
-    KeyMap::SetKeyMapping({Settings::values.pad_sright_key, keyboard_id}, Service::HID::PAD_CIRCLE_RIGHT);
-    KeyMap::SetKeyMapping({Settings::values.pad_sleft_key,  keyboard_id}, Service::HID::PAD_CIRCLE_LEFT);
-    KeyMap::SetKeyMapping({Settings::values.pad_sup_key,    keyboard_id}, Service::HID::PAD_CIRCLE_UP);
-    KeyMap::SetKeyMapping({Settings::values.pad_sdown_key,  keyboard_id}, Service::HID::PAD_CIRCLE_DOWN);
+    for (int i=0; i<Settings::NativeInput::NUM_INPUTS; ++i) {
```
Comment (turbedi) [inline src/citra/emu_window/emu_window_glfw.cpp]: missing space before and after `=` and `<`

File: src/citra_qt/bootmanager.cpp [hunk 5]
```diff
@@ -244,32 +244,9 @@ void GRenderWindow::mouseReleaseEvent(QMouseEvent *event)
...
-
-    KeyMap::SetKeyMapping({Settings::values.pad_cright_key, keyboard_id}, Service::HID::PAD_C_RIGHT);
-    KeyMap::SetKeyMapping({Settings::values.pad_cleft_key,  keyboard_id}, Service::HID::PAD_C_LEFT);
-    KeyMap::SetKeyMapping({Settings::values.pad_cup_key,    keyboard_id}, Service::HID::PAD_C_UP);
-    KeyMap::SetKeyMapping({Settings::values.pad_cdown_key,  keyboard_id}, Service::HID::PAD_C_DOWN);
-    KeyMap::SetKeyMapping({Settings::values.pad_sright_key, keyboard_id}, Service::HID::PAD_CIRCLE_RIGHT);
-    KeyMap::SetKeyMapping({Settings::values.pad_sleft_key,  keyboard_id}, Service::HID::PAD_CIRCLE_LEFT);
-    KeyMap::SetKeyMapping({Settings::values.pad_sup_key,    keyboard_id}, Service::HID::PAD_CIRCLE_UP);
-    KeyMap::SetKeyMapping({Settings::values.pad_sdown_key,  keyboard_id}, Service::HID::PAD_CIRCLE_DOWN);
+    for (int i=0; i<Settings::NativeInput::NUM_INPUTS; ++i) {
```
Comment (turbedi) [inline src/citra_qt/bootmanager.cpp]: missing space before/after = and <

File: src/citra_qt/config.cpp [hunk 6]
```diff
@@ -21,31 +21,21 @@ Config::Config() {
...
-    Settings::values.pad_dright_key = qt_config->value("pad_dright", Qt::Key_H).toInt();
-    Settings::values.pad_sup_key    = qt_config->value("pad_sup",    Qt::Key_Up).toInt();
-    Settings::values.pad_sdown_key  = qt_config->value("pad_sdown",  Qt::Key_Down).toInt();
-    Settings::values.pad_sleft_key  = qt_config->value("pad_sleft",  Qt::Key_Left).toInt();
-    Settings::values.pad_sright_key = qt_config->value("pad_sright", Qt::Key_Right).toInt();
-    Settings::values.pad_cup_key    = qt_config->value("pad_cup",    Qt::Key_I).toInt();
-    Settings::values.pad_cdown_key  = qt_config->value("pad_cdown",  Qt::Key_K).toInt();
-    Settings::values.pad_cleft_key  = qt_config->value("pad_cleft",  Qt::Key_J).toInt();
-    Settings::values.pad_cright_key = qt_config->value("pad_cright", Qt::Key_L).toInt();
+    for (int i=0; i<Settings::NativeInput::NUM_INPUTS; ++i) {
```
Comment (turbedi) [inline src/citra_qt/config.cpp]: missing space before/after = and <

File: src/citra_qt/config.cpp [hunk 7]
```diff
@@ -75,29 +65,9 @@ void Config::ReadValues() {
...
-    qt_config->setValue("pad_sup",    Settings::values.pad_sup_key);
-    qt_config->setValue("pad_sdown",  Settings::values.pad_sdown_key);
-    qt_config->setValue("pad_sleft",  Settings::values.pad_sleft_key);
-    qt_config->setValue("pad_sright", Settings::values.pad_sright_key);
-    qt_config->setValue("pad_cup",    Settings::values.pad_cup_key);
-    qt_config->setValue("pad_cdown",  Settings::values.pad_cdown_key);
-    qt_config->setValue("pad_cleft",  Settings::values.pad_cleft_key);
-    qt_config->setValue("pad_cright", Settings::values.pad_cright_key);
+    for (int i=0; i<Settings::NativeInput::NUM_INPUTS; ++i) {
+        qt_config->setValue(QString::fromStdString(Settings::NativeInput::Mapping[i]), Settings::NativeInput::All[i]);
```
Comment (turbedi) [inline src/citra_qt/config.cpp]: missing space before/after = and <

File: src/citra_qt/config.cpp [hunk 8]
```diff
@@ -75,29 +65,9 @@ void Config::ReadValues() {
...
-    qt_config->setValue("pad_dright", Settings::values.pad_dright_key);
-    qt_config->setValue("pad_sup",    Settings::values.pad_sup_key);
-    qt_config->setValue("pad_sdown",  Settings::values.pad_sdown_key);
-    qt_config->setValue("pad_sleft",  Settings::values.pad_sleft_key);
-    qt_config->setValue("pad_sright", Settings::values.pad_sright_key);
-    qt_config->setValue("pad_cup",    Settings::values.pad_cup_key);
-    qt_config->setValue("pad_cdown",  Settings::values.pad_cdown_key);
-    qt_config->setValue("pad_cleft",  Settings::values.pad_cleft_key);
-    qt_config->setValue("pad_cright", Settings::values.pad_cright_key);
+    for (int i=0; i<Settings::NativeInput::NUM_INPUTS; ++i) {
```
Comment (turbedi) [inline src/citra_qt/config.cpp]: missing space before/after = and <

**Comment by chinhodado** [comment]
Is @jroweboy waiting for more reviews to address them all at once, or is everyone waiting for him to address the current issues first?

(I have interest in seeing this gets merged, since my PR depends on it)

**Comment by jroweboy** [comment]
@chinhodado I went and reworked it to use loops as you recommended, but since then, I've only gotten minor style feedback that I'll go and update soon. I was really hoping to get some Oks from anyone else, and I've made a few mentions about it on IRC, but so far I haven't gotten any reviews yet.

**Comment by chinhodado** [comment]
ping @neobrain

**Comment by kitling** [comment]
My only comment is that almost the same code is being duplicated between the Values enum and the 'All' array. That's probably unavoidable, though.

Other than that and the style stuff, LGTM.

**Comment by jroweboy** [comment]
I've went ahead and corrected all the style issues. Since there are no other comments on it, I suppose this is good to go then?

File: src/core/settings.h [hunk 9]
```diff
@@ -5,34 +5,42 @@
...
+enum Values {
+    A, B, X, Y,
+    L, R, ZL, ZR,
+    START, SELECT, HOME,
+    DUP, DDOWN, DLEFT, DRIGHT,
+    SUP, SDOWN, SLEFT, SRIGHT,
+    CUP, CDOWN, CLEFT, CRIGHT,
+    NUM_INPUTS
+};
+static const std::array<std::string, NUM_INPUTS> Mapping = {
```
Comment (neobrain) [inline src/core/settings.h]: You can just use `const char*` here instead of std::string. Saves some runtime memory allocations.

**Comment by neobrain** [comment]
LGTM other than the string vs const char* thing.

**Comment by jroweboy** [comment]
@neobrain Added that minor change.

**Comment by neobrain** [comment]
Thanks!

**Pull request merged**